driver/src/metrics/http_metrics.rs

```diff
@@ -0,0 +1,197 @@
+use anyhow::Result;
+use prometheus::{HistogramOpts, HistogramVec, Registry};
+use std::sync::Arc;
+use std::time::Duration;
+
+/// A registry for all HTTP related metrics.
+#[derive(Debug)]
+pub struct HttpMetrics {
+    latency: HistogramVec,
+    size: HistogramVec,
+}
+
+impl HttpMetrics {
+    /// Create a new HTTP metrics registry.
+    pub fn new(registry: &Arc<Registry>) -> Result<Self> {
+        let latency = HttpMetrics::initialize_histogram(
+            registry,
+            "dfusion_service_http_latency",
+            "Latency in seconds for HTTP request",
+            vec![0.05, 0.1, 0.25, 0.5, 1.0, 2.0, 5.0, 10.0],
+        )?;
+        let size = HttpMetrics::initialize_histogram(
+            registry,
+            "dfusion_service_http_size",
+            "Size in bytes for HTTP response bodies",
+            vec![
+                1_000.0,
+                10_000.0,
+                100_000.0,
+                1_000_000.0,
+                10_000_000.0,
+                100_000_000.0,
+            ],
+        )?;
+
+        Ok(HttpMetrics { latency, size })
+    }
+
+    /// Initializes a histogram with for all the labels.
+    fn initialize_histogram(
+        registry: &Arc<Registry>,
+        name: &str,
+        description: &str,
+        buckets: Vec<f64>,
+    ) -> Result<HistogramVec> {
+        let options = HistogramOpts::new(name, description).buckets(buckets);
+        let histogram = HistogramVec::new(options, &["client", "request"])?;
+
+        Web3Label::initialize_histogram_labels(&histogram);
+        KrakenLabel::initialize_histogram_labels(&histogram);
+        DexagLabel::initialize_histogram_labels(&histogram);
+        GasStationLabel::initialize_histogram_labels(&histogram);
+
+        registry.register(Box::new(histogram.clone()))?;
+
+        Ok(histogram)
+    }
+
+    /// Add a request latency and size measurement to the current HTTP metrics
+    /// registry for the specified label values.
+    pub fn request<L: LabelValues>(&self, label: L, latency: Duration, size: usize) {
+        self.latency
+            .with_label_values(&label.label_values())
+            .observe(latency.as_secs_f64());
+        self.size
+            .with_label_values(&label.label_values())
+            .observe(size as _);
+    }
+}
+
+impl Default for HttpMetrics {
+    fn default() -> Self {
+        HttpMetrics::new(&Default::default()).unwrap()
+    }
+}
+
+/// A common trait shared between labeled and unlabeled subsystems for getting
+/// the label values to use for metrics.
+pub trait LabelValues {
+    fn label_values(&self) -> [&'static str; 2];
+    fn initialize_histogram_labels(histograms: &HistogramVec);
+}
+
+/// A trait abstracting a labeled subsystem, that is a specific use of an HTTP
+/// client where the HTTP requests should be differenciated.
+pub trait LabeledSubsystem: LabelValues + Sized {
```

Comment:
It took me a while to understand the difference between a Labeled and Unlabelled subsystem. Why do we want to differentiate the two? Couldn't we just log the endpoint even if there is just one query? What's the advantage of doing the differentiation?

Reply:
Its minor, for unlabeled subsystems you can perform HTTP requests without a label (so, `http_client.get_json()` for example), while a labeled subsystem requires a label value (`http_client.get_json(Subsystem::SomeLabel)`).